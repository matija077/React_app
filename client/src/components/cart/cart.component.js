import React from 'react';
import { connect } from 'react-redux';
import {createStructuredSelector} from 'reselect';
import {withRouter} from 'react-router-dom';

import {selectCartItems} from '../../redux/cart/card.selectors';

import SubmitComponent from '../SubmitButton/SubmitButton.component';
import CartItem from '../cart-item/cartItem.component';
import {toggleCartHidden} from '../../redux/cart/cart.actions';
import ClearCart  from '../clear-cart/clearCart.component';

import './cart.style.scss';

function Cart({ cartItems, history, toggleCartHidden }) {
    console.log(cartItems);
    console.log(history);
    console.log(toggleCartHidden);

    return(
        <div className="cart-dropdown">
            <ClearCart></ClearCart>
            <div className="cart-items">
                {
                    cartItems.length ?
                        cartItems.map(cartItem =>
                            <CartItem
                                key={cartItem.id}
                                item={cartItem}>
                            </CartItem>
                        )
                    :
                        <span className="empty-message">
                            Your cart is empty
                        </span>

                }
            </div>
            <SubmitComponent onClick={() => {
                history.push('/checkout');
                toggleCartHidden();
            }}>
                GO TO CHECKOUT
            </SubmitComponent>
        </div>
    );
}

const mapStateToProps = createStructuredSelector({
    cartItemsOld: selectCartItems,
})

export default withRouter(connect(
    mapStateToProps
)(Cart));