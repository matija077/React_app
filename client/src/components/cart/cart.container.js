import React from 'react';
import { Query, Mutation, useMutation, useQuery } from 'react-apollo';

import Cart from './cart.component';

import { queriesAndMutations } from '../../graphQL/resolvers';

function CartContainer({history}) {
    var { TOGGLE_CART_HIDDEN } = queriesAndMutations.mutations;
    var { GET_CART_ITEMS } = queriesAndMutations.queries;

    var { data: { cartItems = [] } = {} } = useQuery(GET_CART_ITEMS);
    var [ toggleCartHidden ] = useMutation(TOGGLE_CART_HIDDEN);

    console.log(cartItems);

    return(
        <Cart
            toggleCartHidden={toggleCartHidden}
            cartItems={cartItems}
            history={history}
        >
        </Cart>
    );
}

export default CartContainer;

/*
<Mutation mutation={TOGGLE_CART_HIDDEN}>
            {
                (mutateFunction = () => {}, {data, error, loading}) => {
                    if (loading || error) {
                        if (error) {
                             console.log(error);
                        }
                        return;
                    }
                    console.log(mutateFunction);

                    return (
                        <Query query={GET_CART_ITEMS}>
                            {
                                ({data, error, loading}) => {
                                    if (loading || error) {
                                        if (error) {
                                             console.log(error);
                                        }
                                        return;
                                    }

                                    console.log(data);

                                    let { cartItems } = data || {};
                                    console.log(cartItems);
                                    return (
                                        <Cart
                                            toggleCartHidden={mutateFunction}
                                            cartItems={cartItems}
                                            history={history}
                                        >
                                        </Cart>
                                    );
                                }
                            }
                        </Query>
                    );
                }
            }
        </Mutation>
*/