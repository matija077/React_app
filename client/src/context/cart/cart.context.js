import { createContext } from 'react';

var CartContext = createContext({
    hidden: true,
    toggleHidden: () => {}
});

export default CartContext;