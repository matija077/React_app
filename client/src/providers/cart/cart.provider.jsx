import React from 'react';
//import { clearCart } from '../../redux/cart/cart.actions';

import {
    removeOneItemFromCart,
    addItemToCart,
    filterItemFromCart,
    getCartItemsCount
} from './cart.utils';

var cartContext = {
    hidden: true,
    cartItemsCount: 0,
    cartItems: [],
    toggleHidden: () => {},
    addItem: () => {},
    removeItem: () => {},
    clearItemFromCart: () => {},
    clearCart: () => {}
};
var CartContext = React.createContext({cartContext});

function CartProvider({ children }) {
    var [hidden, setHidden] = React.useState(cartContext.hidden);
    var [cartItemsCount, setCartItemsCount] = React.useState(
        cartContext.cartItemsCount);
    var [cartItems, setCartItems] = React.useState(
        cartContext.cartItems);

    var toggleCartHidden = function() {
        setHidden(!hidden);
    }
    var addItem = function(item) {
        setCartItems(addItemToCart(cartItems, item));
        /*
        * for debugging
        var newItems = addItemToCart(cartItems, item);
        setCartItems(newItems);
        */
    }
    var removeItem = function(item) {
        setCartItems(removeOneItemFromCart(cartItems, item));
    }
    var clearItemFromCart = function(item) {
        setCartItems(filterItemFromCart(cartItems, item));
    }
    var clearCart = function(){
        console.log("radim");
        setCartItems([]);
    }

    React.useEffect(
        () => {
          setCartItemsCount(getCartItemsCount(cartItems));
        }
        , [cartItems]
    );

    return (
        <CartContext.Provider
            value={{
                hidden,
                cartItemsCount,
                cartItems,
                toggleCartHidden,
                addItem,
                removeItem,
                clearItemFromCart,
                clearCart
            }}
        >
            {children}
        </CartContext.Provider>
    );
}

export { CartContext };
export default CartProvider;