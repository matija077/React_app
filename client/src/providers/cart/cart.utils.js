export function addItemToCart(
    cartItems, cartItemToAdd) {
        const existingCartItem = cartItems.find(
            cartItem => cartItem.id === cartItemToAdd.id
        );

        if (existingCartItem) {
            return cartItems.map(cartItem =>
                cartItem.id === cartItemToAdd.id
                ? {...cartItem, quantity: cartItem.quantity + 1}
                : cartItem
            );
        }

        return [...cartItems, {...cartItemToAdd, quantity: 1}];
}

export function removeOneItemFromCart(cartItems ,cartItemToRemove) {
    const existingCartITemIndex = cartItems.findIndex(
        cartItem => cartItem.id === cartItemToRemove.id
    );
    const existingCartItem = cartItems[existingCartITemIndex];

    console.log(existingCartItem);
    console.log(existingCartITemIndex);
    if (existingCartItem.quantity === 1) {
        cartItems = cartItems.filter(
            cartItem => cartItem.id !== cartItemToRemove.id
        )
    } else {
        cartItems = cartItems.map(
            cartItem => cartItem.id === cartItemToRemove.id ?
                {...cartItem, quantity: cartItem.quantity - 1}
            :
                cartItem
        )
    }

    console.log(cartItems);

    return cartItems;
}

export function filterItemFromCart(cartItems, item) {
    return cartItems.filter(
        cartItem => cartItem.id !== item.id
    )

}

export function getCartItemsCount(cartItems) {
    return cartItems.reduce(
        reducer(quantity)
        , 0
    );
}

function reducer(calculation){
    return function(accumulatedQuantity, item) {
        return accumulatedQuantity + calculation(item);
    }
}

function pricePerQuantity(cartItem) {
    return cartItem.quantity * cartItem.price;
}

function quantity(cartItem) {
    return cartItem.quantity;
}
