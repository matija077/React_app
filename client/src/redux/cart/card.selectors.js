import {createSelector} from 'reselect';

/**
 * inptu selector
 */
function selectCart(state) {
    return state.cart;
}

export var selectCartHidden = createSelector(
    [selectCart],
    (cart) => cart.hidden
);

export var selectCartItems = createSelector(
    [selectCart],
    (cart) => cart.cartItems
);

export var selectCartItemsCount = createSelector(
    [selectCartItems],
    cartItems => cartItems.reduce(
        reducer(quantity)
        ,  0
    )
);

export var selectCartTotal = createSelector(
    [selectCartItems],
     /*cartItems => cartItems.reduce(
        (accumulatedQuantity, cartItem) =>
            accumulatedQuantity + cartItem.prize
        , 0
     )*/
     cartItems => cartItems.reduce(
         reducer(pricePerQuantity)
         , 0
     )
);

function reducer(calculation) {
    return function(accumulatedQuantity, cartItem) {
        return accumulatedQuantity + calculation(cartItem);
    }
}

function pricePerQuantity(cartItem) {
    return cartItem.quantity * cartItem.price;
}

function quantity(cartItem) {
    return cartItem.quantity;
}
