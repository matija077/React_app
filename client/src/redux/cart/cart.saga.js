import React from 'react';
import { all, call, takeLatest, put } from 'redux-saga/effects';

import userActionTypes from '../user/user.types';
import cartActionTypes from '../cart/cart.types';
//import { clearCart } from './cart.actions';
import CartProvider from '../../providers/cart/cart.provider';

function* clearCartFunction() {
    try {
        yield 1;
        //yield put(clearCart());
    } catch(error) {
        console.log("errro clearing cart function");
    }
}

function* onSignOutSuccess() {
    yield takeLatest(
        userActionTypes.SIGN_OUT_SUCCESS,
        clearCartFunction
    )
}

export default function* cartSaga() {
    yield all([
        call(onSignOutSuccess)
    ]);
}