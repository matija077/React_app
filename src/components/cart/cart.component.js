import React from 'react';
import { connect } from 'react-redux';
import {createStructuredSelector} from 'reselect';

import {selectCartItems} from '../../redux/cart/card.selectors';

import SubmitComponent from '../SubmitButton/SubmitButton.component';
import CartItem from '../cart-item/cartItem.component';

import './cart.style.scss';

function Cart({ cartItems }) {
    console.log(cartItems);
    return(
        <div className="cart-dropdown">
            <div className="cart-items">
                {
                    cartItems.length ?
                        cartItems.map(cartItem =>
                            <CartItem
                                key={cartItem.id}
                                item={cartItem}>
                            </CartItem>
                        )
                    :
                        <span className="empty-message">
                            Your cart is empty
                        </span>

                }
            </div>
            <SubmitComponent>GO TO CHECKOUT</SubmitComponent>
        </div>
    );
}

const mapStateToProps = createStructuredSelector({
    cartItems: selectCartItems,
})

export default connect(mapStateToProps)(Cart);