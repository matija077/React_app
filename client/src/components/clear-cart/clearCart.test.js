import React from 'react';
import combineObject from '../../redux/store'
const { store } = combineObject;

import {render as reactRender, fireEvent, screen} from '@testing-library/react'
import { shallow } from 'enzyme';

import { clearCart } from '../../redux/cart/cart.actions';

import ClearCart from './clearCart.component';

describe('test clearCart component', function describe() {
    it("snapshot clearCart enzyme", function snapshotClearCart() {
        expect.assertions(1);
        expect(shallow(<ClearCart store={store} />)).toMatchSnapshot();
    });

    it("clearCart react", function react() {
        expect.assertions(2);
        const hb = "header-block";

        const clearCartComponent = reactRender(<ClearCart
            store={store}
            clearCart={clearCart}
            className={hb}
        >
        </ClearCart>);

        expect(clearCartComponent.container.firstChild.classList.
            contains(hb)).toBe(true);
        expect(screen.getByText("\u274C")).toBeInTheDocument();
    })
});