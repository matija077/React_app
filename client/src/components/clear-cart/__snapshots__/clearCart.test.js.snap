// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test clearCart component snapshot clearCart enzyme 1`] = `ShallowWrapper {}`;
